File: src/rs/src/radio/type_def.rs
use serde::{Deserialize, Serialize};

/// 三维坐标，单位米，格式 (x, y, z)
pub type XYZ = (f64, f64, f64);

/// 链路可达性检查的可选参数
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct RadioCheckOptions {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frequency_mhz: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tx_power_dbm: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rx_sensitivity_dbm: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fresnel_zone_ratio: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub skip_fresnel_zone_check: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_multipath: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_reflection_paths: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_reflection_path_length_ratio: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_tx_antenna_gain_d_bi: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub default_rx_antenna_gain_d_bi: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_antenna_pattern: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_mutual_coupling: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub coupling_negligible_threshold_wavelengths: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_sinr: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub receiver_bandwidth_hz: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_sinr_db: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_frequency_isolation: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_near_field_correction: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub enable_node_body_occlusion: Option<bool>,
}

/// 链路可达性检查请求消息
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckReachabilityRequest {
    pub a_tx: XYZ,
    pub b_rx: XYZ,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<RadioCheckOptions>,
}

/// 更新对象位置的消息
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateObjectPosRequest {
    pub object_id: String,
    pub position: XYZ,
}

/// 更新对象的电磁属性
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateMeshRadioMaterialRequest {
    pub mesh_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub material_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thickness_m: Option<f64>,
}

/// 电磁材料属性定义
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RadioMaterialProperties {
    pub id: String,
    pub display_name: String,
    pub penetration_loss_d_b_per_meter: f64,
    pub reflection_coefficient: f64,
    pub default_thickness_m: f64,
}

/// 摇杆输入消息
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct JoyStickInput {
    pub vx: f64,
    pub vy: f64,
    pub vz: f64,
    pub yaw_rate: f64,
}

File: src/lua/radio/type_def.lua
-- Radio related type definitions in Lua

local type_def = {}

-- Helper functions
local function xyz_from_list(data)
    return { tonumber(data[1]), tonumber(data[2]), tonumber(data[3]) }
end

-- ---------------------------------------------------------------------------
-- RadioCheckOptions
-- ---------------------------------------------------------------------------
local RadioCheckOptions = {}
RadioCheckOptions.__index = RadioCheckOptions

function RadioCheckOptions.new(data)
    local self = setmetatable({}, RadioCheckOptions)
    data = data or {}
    self.frequencyMHz = data.frequencyMHz
    self.txPowerDbm = data.txPowerDbm
    self.rxSensitivityDbm = data.rxSensitivityDbm
    self.fresnelZoneRatio = data.fresnelZoneRatio
    self.skipFresnelZoneCheck = data.skipFresnelZoneCheck
    self.enableMultipath = data.enableMultipath
    self.maxReflectionPaths = data.maxReflectionPaths
    self.maxReflectionPathLengthRatio = data.maxReflectionPathLengthRatio
    self.defaultTxAntennaGain_dBi = data.defaultTxAntennaGain_dBi
    self.defaultRxAntennaGain_dBi = data.defaultRxAntennaGain_dBi
    self.enableAntennaPattern = data.enableAntennaPattern
    self.enableMutualCoupling = data.enableMutualCoupling
    self.couplingNegligibleThresholdWavelengths = data.couplingNegligibleThresholdWavelengths
    self.enableSINR = data.enableSINR
    self.receiverBandwidthHz = data.receiverBandwidthHz
    self.minSINR_dB = data.minSINR_dB
    self.enableFrequencyIsolation = data.enableFrequencyIsolation
    self.enableNearFieldCorrection = data.enableNearFieldCorrection
    self.enableNodeBodyOcclusion = data.enableNodeBodyOcclusion
    return self
end

function RadioCheckOptions:to_dict()
    local d = {}
    for k, v in pairs(self) do
        if v ~= nil then
            d[k] = v
        end
    end
    return d
end

-- ---------------------------------------------------------------------------
-- CheckReachabilityRequest
-- ---------------------------------------------------------------------------
local CheckReachabilityRequest = {}
CheckReachabilityRequest.__index = CheckReachabilityRequest

function CheckReachabilityRequest.new(aTx, bRx, options)
    local self = setmetatable({}, CheckReachabilityRequest)
    self.aTx = aTx
    self.bRx = bRx
    self.options = options
    return self
end

function CheckReachabilityRequest:to_dict()
    local d = {
        aTx = self.aTx,
        bRx = self.bRx
    }
    if self.options then
        d.options = self.options:to_dict()
    end
    return d
end

-- ---------------------------------------------------------------------------
-- UpdateObjectPosRequest
-- ---------------------------------------------------------------------------
local UpdateObjectPosRequest = {}
UpdateObjectPosRequest.__index = UpdateObjectPosRequest

function UpdateObjectPosRequest.new(objectId, position)
    local self = setmetatable({}, UpdateObjectPosRequest)
    self.objectId = objectId
    self.position = position
    return self
end

function UpdateObjectPosRequest:to_dict()
    return {
        objectId = self.objectId,
        position = self.position
    }
end

-- ---------------------------------------------------------------------------
-- UpdateMeshRadioMaterialRequest
-- ---------------------------------------------------------------------------
local UpdateMeshRadioMaterialRequest = {}
UpdateMeshRadioMaterialRequest.__index = UpdateMeshRadioMaterialRequest

function UpdateMeshRadioMaterialRequest.new(meshId, materialId, thickness_m)
    local self = setmetatable({}, UpdateMeshRadioMaterialRequest)
    self.meshId = meshId
    self.materialId = materialId
    self.thickness_m = thickness_m
    return self
end

function UpdateMeshRadioMaterialRequest:to_dict()
    local d = { meshId = self.meshId }
    if self.materialId then d.materialId = self.materialId end
    if self.thickness_m then d.thickness_m = self.thickness_m end
    return d
end

-- ---------------------------------------------------------------------------
-- RadioMaterialProperties
-- ---------------------------------------------------------------------------
local RadioMaterialProperties = {}
RadioMaterialProperties.__index = RadioMaterialProperties

function RadioMaterialProperties.new(data)
    local self = setmetatable({}, RadioMaterialProperties)
    self.id = data.id
    self.displayName = data.displayName
    self.penetrationLoss_dBPerMeter = data.penetrationLoss_dBPerMeter
    self.reflectionCoefficient = data.reflectionCoefficient
    self.defaultThickness_m = data.defaultThickness_m
    return self
end

function RadioMaterialProperties:to_dict()
    return {
        id = self.id,
        displayName = self.displayName,
        penetrationLoss_dBPerMeter = self.penetrationLoss_dBPerMeter,
        reflectionCoefficient = self.reflectionCoefficient,
        defaultThickness_m = self.defaultThickness_m
    }
end

-- ---------------------------------------------------------------------------
-- JoyStickInput
-- ---------------------------------------------------------------------------
local JoyStickInput = {}
JoyStickInput.__index = JoyStickInput

function JoyStickInput.new(data)
    local self = setmetatable({}, JoyStickInput)
    data = data or {}
    self.vx = data.vx or 0.0
    self.vy = data.vy or 0.0
    self.vz = data.vz or 0.0
    self.yawRate = data.yawRate or 0.0
    return self
end

function JoyStickInput:to_dict()
    return {
        vx = self.vx,
        vy = self.vy,
        vz = self.vz,
        yawRate = self.yawRate
    }
end

-- Module exports
type_def.RadioCheckOptions = RadioCheckOptions
type_def.CheckReachabilityRequest = CheckReachabilityRequest
type_def.UpdateObjectPosRequest = UpdateObjectPosRequest
type_def.UpdateMeshRadioMaterialRequest = UpdateMeshRadioMaterialRequest
type_def.RadioMaterialProperties = RadioMaterialProperties
type_def.JoyStickInput = JoyStickInput

function type_def.radio_check_options_from_dict(data)
    return RadioCheckOptions.new(data)
end

function type_def.check_reachability_request_from_dict(data)
    local opts = nil
    if data.options then
        opts = RadioCheckOptions.new(data.options)
    end
    return CheckReachabilityRequest.new(xyz_from_list(data.aTx), xyz_from_list(data.bRx), opts)
end

function type_def.update_object_pos_request_from_dict(data)
    return UpdateObjectPosRequest.new(tostring(data.objectId), xyz_from_list(data.position))
end

function type_def.update_mesh_radio_material_request_from_dict(data)
    return UpdateMeshRadioMaterialRequest.new(tostring(data.meshId), data.materialId, data.thickness_m)
end

function type_def.radio_material_properties_from_dict(data)
    return RadioMaterialProperties.new(data)
end

function type_def.joystick_input_from_dict(data)
    return JoyStickInput.new(data)
end

return type_def
